### PR TITLE
[XPU] reuse logits and softmax to avoid redundant memory allocation

### PR DESCRIPTION
**Pull request by runzhech**
### PR Category
Custom Device

### PR Types
Improvements

### Description
Using "ShareDataWith" instead of  "TensorCopy" for logits and softmax to avoid redundant memory alloc.

### CODE REVIEW

**Review by zhiqiu — DISMISSED** [review]
LGTM for shareDataWith

File: paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op_xpu.cc [hunk 1]
```diff
@@ -152,18 +152,13 @@ struct CSoftmaxWithCrossEntropyProcessGroupFunctor<phi::XPUContext, T> {
...
 
     phi::DenseTensor logits_2d, softmax_2d;
-    framework::TensorCopy(
-        *logits, ctx.GetPlace(), ctx.device_context(), &logits_2d);
-    framework::TensorCopy(
-        *softmax, ctx.GetPlace(), ctx.device_context(), &softmax_2d);
-    logits_2d.Resize({N, D});
-    softmax_2d.Resize({N, D});
+    logits_2d.ShareDataWith(*logits).Resize({N, D});
+    softmax_2d.ShareDataWith(*softmax).Resize({N, D});
```
Comment (lj970926) [inline paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op_xpu.cc:161]: 下面CSoftmaxWithCrossEntropyFunctor（line 389~line 395）那里是不是也可以一起改下？

**Comment by houj04** [comment]
这个PR修改了`CSoftmaxWithCrossEntropyProcessGroupFunctor`里面的相关代码，那`CSoftmaxWithCrossEntropyFunctor`里面也有类似的代码，需要改吗？

**Review by lj970926 — APPROVED** [review]
LGTM

File: paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op_xpu.cc (hunk 1 above)
Comment (HarperCy) [inline paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op_xpu.cc:161]: sharedatawith我记得paddle同学说过，是访问同一块地址，会有潜在的风险。之前paddle cast，在数据类型相同时，就是用了sharedatawith，模型训练有时就会有奇怪的现象

Comment (runzhech) [reply paddle/fluid/operators/collective/c_softmax_with_cross_entropy_op_xpu.cc:161]: 这个算子比较特殊，sharedatawith的是两个临时tensor。其中logits_2d不会被修改，softmax_2d在修改后原本就会被写回原地址，逻辑上跟之前一样，所以不会有问题。

**Review by zhiqiu — APPROVED** [review]
LGTM

**Review by lj970926 — APPROVED** [review]
LGTM

**Review by RuohengMa — APPROVED** [review]
LGTM

**Review by houj04 — APPROVED** [review]
